### PR TITLE
Refactor ServiceProxy object to enable testing

### LINKED ISSUE #1791: Modify ServiceProxy object to use abstraction created by issue #1673
> estimate 1

Parent issue: #1772

Done criteria: Reimplement the logic in ServiceProxy.cpp to use the sd_bus interface abstraction.

### PR DESCRIPTION
**Pull request by cmcantalupo**
- Resloves #1673 
- Resolves #1791

### CODE REVIEW

File: src/SDBus.hpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+namespace geopm
+{
+    class SDBusMessage;
+
+    class SDBus
+    {
+        public:
+            SDBus() = default;
+            virtual ~SDBus() = default;
+            static std::unique_ptr<SDBus> make_unique(void);
```
Comment (cmcantalupo) [inline src/SDBus.hpp:65]: Doxygen missing

File: src/SDBusMessage.hpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+namespace geopm
+{
+    class SDBusMessage
+    {
+        public:
+            enum m_message_type_e {
+                M_MESSAGE_TYPE_STRUCT,
+                M_MESSAGE_TYPE_ARRAY,
+            };
+            SDBusMessage() = default;
```
Comment (cmcantalupo) [inline src/SDBusMessage.hpp:62]: Doxygen missing

File: src/SDBus.hpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                int arg1,
+                int arg2,
+                double arg3) override;
+            std::shared_ptr<SDBusMessage> make_call_message(
+                const std::string &member) override;
+        private:
+            sd_bus *m_bus;
+            const char * m_dbus_destination;
+            const char * m_dbus_path;
+            const char * m_dbus_interface;
```
Comment (bgeltz) [inline src/SDBus.hpp]:
```suggestion
            const char *m_dbus_interface;
```

File: src/SDBus.cpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "SDBus.hpp"
+
+#include <sstream>
```
Comment (bgeltz) [inline src/SDBus.cpp:35]:
```suggestion
#include <sstream>
#include <systemd/sd-bus.h>
```

File: src/SDBusMessage.cpp [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        int ret = sd_bus_message_read(m_bus_message, "d", &result);
+        check_bus_error("sd_bus_message_read", ret);
+        m_was_success = (ret != 0);
+        return result;
+    }
+
+    int SDBusMessageImp::read_integer(void)
+    {
+        check_null_ptr(__func__, m_bus_message);
+        int result = -1;
```
Comment (bgeltz) [inline src/SDBusMessage.cpp]: We need some sort of number that's not valid (or at least very rare like MAX_INT) here.

**Review by bgeltz — APPROVED** [review]

**Pull request merged**